aten/src/ATen/native/cpu/DistributionKernels.cpp

```diff
@@ -103,11 +103,88 @@ void bernoulli_scalar_kernel(const TensorBase &self, double p, c10::optional<Gen
 }
 
+#if !AT_MKL_ENABLED()
+void exponential_kernel(TensorIteratorBase& iter, double lambda, c10::optional<Generator> gen) {
+  exponential_kernel_default(iter, lambda, gen);
+}
+#else
+void exponential_kernel(TensorIteratorBase &iter, double lambda, c10::optional<Generator> gen) {
+  Tensor self = iter.tensor(0);
+  if (lambda > 0 && !std::isinf(lambda) && !std::isnan(lambda) && cpuinfo_initialize() &&
+      cpuinfo_vendor_intel == cpuinfo_get_processor(0)->core->vendor) {
+    CPUGeneratorImpl* generator = get_generator_or_default<CPUGeneratorImpl>(gen, detail::getDefaultCPUGenerator());
+    int64_t seed;
+    {
+      // See Note [Acquire lock when using random generators]
+      std::lock_guard<std::mutex> lock(generator->mutex_);
+      if (self.scalar_type() == at::kDouble)
+        seed = generator->random64();
+      else
+        seed = generator->random();
+    }
+    int64_t n = self.numel();
+    bool contig = self.is_contiguous();
+
+    AT_DISPATCH_ALL_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, self.scalar_type(), "exponential_cpu", [&] {
+      at::Tensor tmp_tensor;
+      constexpr bool is_df = std::is_same<scalar_t, float>::value || std::is_same<scalar_t, double>::value;
+      if (is_df && contig) {
+        tmp_tensor = self;
+      } else if (std::is_same<scalar_t, double>::value) {
+        tmp_tensor = at::empty(self.sizes(), self.options().dtype(at::kDouble));
+      } else {
+        tmp_tensor = at::empty(self.sizes(), self.options().dtype(at::kFloat));
+      }
+
+      scalar_t *self_ptr = self.data_ptr<scalar_t>();
+      using tmp_scalar_t = typename std::conditional_t<std::is_same<scalar_t, double>::value, double, float>;
+      tmp_scalar_t *sample_ptr = tmp_tensor.data_ptr<tmp_scalar_t>();
+
+      auto sample = [&](int64_t begin, int64_t end) {
+        int64_t len = end - begin;
+        if (len > 0) {
+          VSLStreamStatePtr stream;
+          if (std::is_same<scalar_t, double>::value) {
+            vslNewStream(&stream, VSL_BRNG_MCG31, seed);
+            vslSkipAheadStream(stream, begin);
+            vdRngExponential(VSL_RNG_METHOD_EXPONENTIAL_ICDF, stream, len,
+              (double *)(sample_ptr + begin), 0, 1./lambda);
+            vslDeleteStream(&stream);
+          } else {
+            vslNewStream(&stream, VSL_BRNG_MCG31, seed);
+            vslSkipAheadStream(stream, begin);
+            vsRngExponential(VSL_RNG_METHOD_EXPONENTIAL_ICDF, stream, len,
+              (float *) (sample_ptr + begin), 0, 1./lambda);
+            vslDeleteStream(&stream);
+          }
+          // vectorized copy if using buffer and contiguous
+          if (!is_df && contig) {
+            scalar_t *self_seg = self_ptr + begin;
+            tmp_scalar_t *tmp_seg = sample_ptr + begin;
+            at::vec::convert<tmp_scalar_t, scalar_t>(tmp_seg, self_seg, len);
+          }
+        }
+      };
+
+      parallel_for(0, n, /* grain_size= */ 800, sample);
```

Comment:
Generally looks good to me. Does the grain_size is fixed number for any length?

Reply:
Yes, Generally, 800 is a heuristic grain_size with relatively good performance.